Mathlib/CategoryTheory/Limits/FormalCoproducts.lean

```diff
@@ -0,0 +1,426 @@
+    (h₂ : ∀ (i : X.I), f.φ i ≫ eqToHom (by rw [h₁]) = g.φ i) : f = g := by
+  obtain ⟨f, F⟩ := f
+  obtain ⟨g, G⟩ := g
+  obtain rfl : f = g := h₁
+  obtain rfl : F = G := by ext i; simpa using h₂ i
+  rfl
+
+lemma hom_ext_iff {X Y : FormalCoproduct.{w} C} (f g : X ⟶ Y) :
+    f = g ↔ ∃ h₁ : f.f = g.f, ∀ (i : X.I), f.φ i ≫ eqToHom (by rw [h₁]) = g.φ i :=
+  ⟨(· ▸ by simp), fun ⟨h₁, h₂⟩ ↦ hom_ext h₁ h₂⟩
+
+lemma hom_ext_iff' {X Y : FormalCoproduct.{w} C} (f g : X ⟶ Y) :
+    f = g ↔ ∀ i : X.I, ∃ h₁ : f.f i = g.f i, f.φ i ≫ eqToHom (by rw [h₁]) = g.φ i :=
+  ⟨(· ▸ by simp), fun h ↦ hom_ext (funext fun i ↦ (h i).fst) fun i ↦ (h i).snd⟩
+
+/-- A way to create isomorphisms in the category of formal coproducts, by creating an `Equiv`
+between the indexing sets, and then correspondingly isomorphisms of each component. -/
+@[simps!] def isoOfComponents {X Y : FormalCoproduct.{w} C} (e : X.I ≃ Y.I)
+    (h : ∀ i, X.obj i ≅ Y.obj (e i)) : X ≅ Y where
+  hom := { f := e, φ := fun i ↦ (h i).hom }
+  inv := { f := e.symm, φ := fun i ↦ eqToHom (by simp) ≫ (h (e.symm i)).inv }
+  hom_inv_id := by ext <;> aesop
+  inv_hom_id := by ext <;> aesop
+
+variable (C) in
+/-- An object of the original category produces a formal coproduct on that object only, so indexed
+by `PUnit`, the type with one element. -/
+@[simps!] def of : C ⥤ FormalCoproduct.{w} C where
+  obj X := ⟨PUnit, fun _ ↦ X⟩
+  map f := ⟨fun _ ↦ PUnit.unit, fun _ ↦ f⟩
+
+section ofHom
+
+variable {X : C} {Y : FormalCoproduct.{w} C}
+
+/-- A map `of(X) ⟶ Y` is specified by an element of `Y`'s indexing set, and then a morphism
+`X ⟶ Y.obj i` in the original category. -/
+def ofHom.mk (i : Y.I) (f : X ⟶ Y.obj i) : (of C).obj X ⟶ Y :=
+  ⟨fun _ ↦ i, fun _ ↦ f⟩
+
+/-- A map `of(X) ⟶ Y` is specified by an element of `Y`'s indexing set, and then a morphism
+`X ⟶ Y.obj i` in the original category. -/
+def ofHom.fst (f : (of C).obj X ⟶ Y) : Y.I :=
+  f.f PUnit.unit
+
+/-- A map `of(X) ⟶ Y` is specified by an element of `Y`'s indexing set, and then a morphism
+`X ⟶ Y.obj i` in the original category. -/
+def ofHom.snd (f : (of C).obj X ⟶ Y) :
+    X ⟶ Y.obj (ofHom.fst f) :=
+  f.φ PUnit.unit
+
+lemma ofHom.mk_fst_snd (f : (of C).obj X ⟶ Y) : ofHom.mk (ofHom.fst f) (ofHom.snd f) = f := by
+  ext <;> aesop
+
+end ofHom
+
+-- This is probably some form of adjunction?
+/-- A map `of(X) ⟶ Y` is specified by an element of `Y`'s indexing set, and then a morphism
+`X ⟶ Y.obj i` in the original category. -/
+def ofHomEquiv (X : C) (Y : FormalCoproduct.{w} C) :
+    ((of C).obj X ⟶ Y) ≃ (i : Y.I) × (X ⟶ Y.obj i) where
+  toFun f := ⟨ofHom.fst f, ofHom.snd f⟩
+  invFun f := ofHom.mk f.1 f.2
+  left_inv f := ofHom.mk_fst_snd f
+  right_inv _ := rfl
+
+/-- `of` is fully faithful, which means that `(X ⟶ Y) ≃ (of(X) ⟶ of(Y))`. -/
+def fullyFaithfulOf : (of C).FullyFaithful where
+  preimage f := f.φ PUnit.unit
+
+instance : (of C).Full :=
+  fullyFaithfulOf.full
+
+instance : (of C).Faithful :=
+  fullyFaithfulOf.faithful
+
+
```

Comment:
```suggestion
```
A bit of style nitpicking here, but I don’t see many file in the library that puts 2 blank lines before sections. There are a bunch of other cases below.

Reply:
I think extra newlines are fine.